messaging/src/main/java/io/lantern/messaging/Messaging.kt

```diff
@@ -0,0 +1,849 @@
+        val storedAttachment =
+            newStoredAttachment.setAttachment(attachment.build())
+        encryptAttachment(storedAttachment, inputStream, length)
+        return storedAttachment.build()
+    }
+
+    /**
+     * Encrypts the given attachment and stores it to disk.
+     *
+     * @param storedAttachment the attachment to encrypt
+     * @param inputStream source of data to encrypt (if attachment wasn't created from a file)
+     * @param length length of data in inputStream
+     */
+    fun encryptAttachment(
+        storedAttachment: Model.StoredAttachment.Builder,
+        inputStream: InputStream? = null,
+        length: Long? = null
+    ) {
+        val plainTextFile = File(storedAttachment.plainTextFilePath)
+        if (inputStream == null && !plainTextFile.exists()) {
+            throw AttachmentPlainTextMissingException()
+        }
+        val attachmentBuilder = storedAttachment.attachment.toBuilder()
+        encryptAttachment(
+            attachmentBuilder,
+            inputStream ?: FileInputStream(plainTextFile),
+            FileOutputStream(storedAttachment.encryptedFilePath),
+            length ?: plainTextFile.length()
+        )
+        storedAttachment.attachment = attachmentBuilder.build()
+        storedAttachment.status = Model.StoredAttachment.Status.PENDING_UPLOAD
+    }
+
+    private fun encryptAttachment(
+        attachmentBuilder: Model.Attachment.Builder,
+        inputStream: InputStream,
+        outputStream: OutputStream,
+        length: Long
+    ) {
+        inputStream.use { input ->
+            val maxLength = cfg.get().maxAttachmentSize
+            if (AttachmentCipherOutputStream.getCiphertextLength(length) > maxLength) {
+                val maxPlainTextLength =
+                    maxLength - AttachmentCipherOutputStream.MAXIMUM_ENCRYPTION_OVERHEAD
+                throw AttachmentTooBigException(maxPlainTextLength)
+            }
+
+            val keyMaterial = ByteArray(64)
+            SecureRandom().nextBytes(keyMaterial)
+            val output =
+                AttachmentCipherOutputStream(
+                    keyMaterial,
+                    outputStream
+                )
+            val plaintextLength = Util.copy(input, output)
+            attachmentBuilder.keyMaterial = keyMaterial.byteString()
+            attachmentBuilder.plaintextLength = plaintextLength
+            attachmentBuilder.digest = output.transmittedDigest.byteString()
+        }
+    }
+
+    internal val newStoredAttachment: Model.StoredAttachment.Builder
+        get() {
+            val guid = UUID.randomUUID().toString()
+            // break attachmentsDirectory into smaller subfolders to avoid having too many files in any one folder
+            val subDirectory = File(
+                arrayOf(
+                    attachmentsDirectory.absolutePath,
+                    guid.substring(0, 1),
+                    guid.substring(1, 2),
+                    guid.substring(2, 3)
```

Comment:
Is this different than the following?
```
attachmentsDirectory.absolutePath,
guid.substring(0, 3),
```

Reply:
Yes.

Let's say the GUID is `1234567890`.

The existing code produces the path `/attachments/1/2/3` whereas your suggestion produces `/attachments/123`. The distinction matters because one case results in fewer files per folder, which in depending on file system can have a performance impact if the number of files gets really large.

Reply:
Ahh I see.  Thanks!